onnx/onnx.in.proto

```diff
@@ -417,12 +514,47 @@ message TypeProto {
 
 // #if ONNX-ML
 
-  // repeated T
-  message Sequence {
-  // The type and optional shape of each element of the sequence.
-  // This field MUST be present for this version of the IR.
-    optional TypeProto elem_type = 1;
-  };
+  message Scalar { 
+    // This field MUST NOT have the value of UNDEFINED 
+    // This field MUST be present for this version of the IR. 
+    optional TensorProto.DataType data_type = 1; 
+  }; 
+
+  message Sequence { 
+    // The type and optional shape of each element of the sequence. 
+    // This field MUST be present for this version of the IR. 
+    optional TypeProto elem_type = 1; 
+  }; 
+
+  message Tuple { 
+    // The element types of the tuple. 
+    repeated TypeProto elem_type = 1; 
+  }; 
+
+  message Record { 
+    // The type and shape of each field is described by a ValueInfoProto. 
+    // The field names must be unique. 
+    // This field MUST be present for this version of the IR. 
+    repeated ValueInfoProto fields = 1; 
+  }; 
+
+  message Opaque { 
+    // Domain with empty string means ONNX domain. 
+    optional string domain = 1; 
+    // The name is optional but significant when provided. 
+    optional string name = 2; 
+  }; 
+
+  message SparseTensor { 
```

Comment:
Are there cases where it is worthwhile for the model to indicate the specific sparse representation to use? Or, is it sufficient to leave it to the backend/implementation to make the decision?

Reply:
I'm not sure about that. It should be an abstraction, I think. There's a case to be made for having separate representations for tensors that will not be updated (such as attributes) and those that will (such as those whose allocated space is reused by operators), but that seems pretty subtle.